crates/jstzd/src/image.rs

```diff
@@ -0,0 +1,160 @@
+        }
+        create_image(self, client.clone()).await
+    }
+}
+
+pub struct DockerImage<'a> {
+    image_name: &'a str,
+    image_tag: Option<&'a str>,
+    entrypoint: Option<&'a str>,
+    exposed_ports: Option<&'a [u16]>,
+}
+
+impl Image for DockerImage<'_> {
+    fn image_name(&self) -> &str {
+        self.image_name
+    }
+
+    fn entrypoint(&self) -> Option<&str> {
+        self.entrypoint
+    }
+
+    fn image_tag(&self) -> &str {
+        self.image_tag.unwrap_or("latest")
+    }
+
+    fn exposed_ports(&self) -> &[u16] {
+        self.exposed_ports.unwrap_or(&[])
+    }
+}
+
+pub struct DockerImageBuilder<'a> {
+    image_name: &'a str,
+    image_tag: Option<&'a str>,
+    entrypoint: Option<&'a str>,
+    exposed_ports: Option<&'a [u16]>,
+}
+
+impl<'a> DockerImageBuilder<'a> {
+    pub fn new(image_name: &'a str) -> DockerImageBuilder<'a> {
+        DockerImageBuilder {
+            image_name,
+            image_tag: None,
+            entrypoint: None,
+            exposed_ports: None,
+        }
+    }
+
+    pub fn with_tag(mut self, image_tag: &'a str) -> Self {
+        self.image_tag = Some(image_tag);
+        self
+    }
+
+    pub fn with_entrypoint(mut self, entrypoint: &'a str) -> Self {
+        self.entrypoint = Some(entrypoint);
+        self
+    }
+
+    pub fn with_exposed_ports(mut self, exposed_ports: &'a [u16]) -> Self {
+        self.exposed_ports = Some(exposed_ports);
+        self
+    }
+
+    pub fn build(self) -> DockerImage<'a> {
+        DockerImage {
+            image_name: self.image_name,
+            image_tag: self.image_tag,
+            entrypoint: self.entrypoint,
+            exposed_ports: self.exposed_ports,
+        }
+    }
+}
+
+async fn image_exists<I>(image: &I, client: Arc<Docker>) -> bool
+where
+    I: Image + ?Sized,
+{
+    let reference = format!("{}:{}", image.image_name(), image.image_tag());
+    let filters = [("reference".to_string(), vec![reference])]
+        .into_iter()
+        .collect::<HashMap<_, _>>();
+    let images = &client
+        .list_images(Some(ListImagesOptions::<String> {
+            all: true,
+            filters,
+            ..Default::default()
+        }))
+        .await;
+    match images {
+        Ok(images) => !images.is_empty(),
+        Err(_) => false,
+    }
+}
+
+async fn create_image<I>(image: &I, client: Arc<Docker>) -> anyhow::Result<()>
+where
+    I: Image + ?Sized,
+{
+    let options = Some(CreateImageOptions {
+        from_image: image.image_name(),
+        tag: image.image_tag(),
+        ..Default::default()
+    });
+    let mut stream = client.create_image(options, None, None);
+    while let Some(create_info) = stream.next().await {
+        match create_info {
+            Ok(info) => {
+                if let Some(status) = info.status {
+                    println!("{:?}", status)
+                }
+            }
+            Err(e) => return Err(e.into()),
+        }
+    }
+    Ok(())
+}
```

Comment:
Suggest that these are methods on `RunnableImage`

Reply:
i'm not sure if we should instantiate a `RunnableImage` with a container name inside `pull_image`. do you mean associated function rather than a method?

Reply:
put them as a default method on `Image` trait